### PR TITLE
Expose tensors passed with the query in IRequestContext.

### PR DESCRIPTION
**Pull request by geirst**
@havardpe please review
@arnej27959 FYI

### CODE REVIEW

File: searchcore/src/tests/proton/matching/request_context/request_context_test.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+public:
+    RequestContextTest()
+        : _clock(),
+          _doom(_clock, fastos::SteadyTimeStamp()),
+          _attr_ctx(),
+          _props(),
+          _request_ctx(_doom, _attr_ctx, _props),
+          _query_tensor(makeTensor<Tensor>(TensorSpec("tensor(x[2])")
```
Comment (havardpe) [inline searchcore/src/tests/proton/matching/request_context/request_context_test.cpp]: prefer making tensors (values) using the tensor engine from_spec function

File: searchcore/src/tests/proton/matching/request_context/request_context_test.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                                                   .add({{"x", 0}}, 3).add({{"x", 1}}, 5)))
+    {
+        insert_tensor_in_properties("my_tensor", *_query_tensor);
```
Comment (havardpe) [inline searchcore/src/tests/proton/matching/request_context/request_context_test.cpp]: consider using tensor engine encode function directly

File: searchcore/src/tests/proton/matching/request_context/request_context_test.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
+// Copyright 2019 Oath Inc. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include <vespa/eval/eval/tensor_spec.h>
+#include <vespa/eval/tensor/serialization/typed_binary_format.h>
+#include <vespa/eval/tensor/tensor.h>
+#include <vespa/eval/tensor/test/test_utils.h>
```
Comment (havardpe) [inline searchcore/src/tests/proton/matching/request_context/request_context_test.cpp]: I think we should try to deprecate this one in favor of the eval/eval/test library and using tensor engine more directly

File: searchcore/src/tests/proton/matching/request_context/request_context_test.cpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    Tensor::UP get_query_tensor(const vespalib::string& tensor_name) const {
+        return _request_ctx.get_query_tensor(tensor_name);
+    }
+};
+
+TEST_F(RequestContextTest, query_tensor_can_be_retrieved)
+{
+    auto tensor = get_query_tensor("my_tensor");
+    ASSERT_TRUE(tensor);
+    EXPECT_EQUAL(expected_query_tensor(), *tensor);
```
Comment (havardpe) [inline searchcore/src/tests/proton/matching/request_context/request_context_test.cpp]: prefer comparing TensorSpec objects (tensor engine to_spec)

File: searchcore/src/vespa/searchcore/proton/matching/requestcontext.cpp [hunk 5]
```diff
@@ -23,8 +34,27 @@ RequestContext::getAttributeStableEnum(const vespalib::string &name) const
...
 }
 
-void RequestContext::asyncForAttribute(const vespalib::string &name, std::unique_ptr<IAttributeFunctor> func) const {
+void
+RequestContext::asyncForAttribute(const vespalib::string &name, std::unique_ptr<IAttributeFunctor> func) const
+{
     _attributeContext.asyncForAttribute(name, std::move(func));
 }
 
+vespalib::tensor::Tensor::UP
```
Comment (havardpe) [inline searchcore/src/vespa/searchcore/proton/matching/requestcontext.cpp]: I think this should return eval::Value::UP instead

File: searchcore/src/vespa/searchcore/proton/matching/requestcontext.cpp [hunk 6]
```diff
@@ -23,8 +34,27 @@ RequestContext::getAttributeStableEnum(const vespalib::string &name) const
...
+RequestContext::get_query_tensor(const vespalib::string& tensor_name) const
+{
+    auto property = _rank_properties.lookup(tensor_name);
+    if (property.found() && !property.get().empty()) {
+        const vespalib::string& value = property.get();
+        vespalib::nbostream stream(value.data(), value.size());
+        try {
+            return vespalib::tensor::TypedBinaryFormat::deserialize(stream);
```
Comment (havardpe) [inline searchcore/src/vespa/searchcore/proton/matching/requestcontext.cpp]: prefer (default) tensor engine decode here

**Review by havardpe — COMMENTED** [review]
please consider my comments

**Comment by geirst** [comment]
@havardpe PTAL

**Review by havardpe — APPROVED** [review]

**Pull request merged**